### PR TITLE
Example: Simple RL example using DQN/Lightning

### LINKED ISSUE #713: Support for non-static data for reinforcement learning
What would be the best approach for reinforcement learning problems where you would need to interact with the environment for data? Maybe DataLoader is restricting?

### PR DESCRIPTION
**Pull request by djbyrne**
DQN RL Agent using Lightning. Model uses an IterableDataset to wrap the ReplayBuffer, providing mini batches of past experiences to train on during each train_step. During each train_step, the agent carries out a step through the environment and updates the ReplayBuffer within the Dataset.

# Before submitting

- [X] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [X] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure to update the docs?   
- [ ] Did you write any new necessary tests?  
- [x] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CHANGELOG.md)?

## What does this PR do?
Fixes #713
Provides a basic domain example of using Lightning for Reinforcement Learning

## PR review    
Anyone in the community is free to review the PR once the tests have passed.     
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: pl_examples/domain_templates/dqn.py [hunk 1]
```diff
@@ -0,0 +1,271 @@
...
+import gym
+import numpy as np
+import collections
+
+import torch
+import torch.nn as nn
+import torch.optim as optim
+from torch.optim import Optimizer
+from torch.utils.data import DataLoader
+from torch.utils.data.dataset import IterableDataset
```
Comment (Borda) [inline pl_examples/domain_templates/dqn.py]:
```suggestion
import argparse
import collections

import gym
import numpy as np
import torch
import torch.nn as nn
import torch.optim as optim
from torch.optim import Optimizer
from torch.utils.data import DataLoader
from torch.utils.data.dataset import IterableDataset

import pytorch_lightning as pl
```
add the gym to pl_examples/requitements.txt

File: pl_examples/domain_templates/dqn.py [hunk 2]
```diff
@@ -0,0 +1,271 @@
...
+
+
+class DQN(nn.Module):
+    """ Simple MLP network"""
+
+    def __init__(self, obs_size, n_actions, hidden_size=128):
```
Comment (Borda) [inline pl_examples/domain_templates/dqn.py]: pls add types

File: pl_examples/domain_templates/dqn.py [hunk 3]
```diff
@@ -0,0 +1,271 @@
...
+        return self.__dataloader()
+
+    def get_device(self, batch) -> str:
+        """Retrieve device currently being used by minibatch"""
+        if self.on_gpu:
+            device = batch[0].device.index
+        else:
+            device = 'cpu'
+
+        return device
```
Comment (Borda) [inline pl_examples/domain_templates/dqn.py]:
```suggestion
        return batch[0].device.index if self.on_gpu else 'cpu'
```

**Comment by williamFalcon** [comment]
merged #990 
can you also transfer this to a colab?

File: pl_examples/domain_templates/dqn.py (hunk 2 above)
Comment (djbyrne) [reply pl_examples/domain_templates/dqn.py]: Done

**Comment by djbyrne** [comment]
@williamFalcon how do I go about adding the example to the colab notebook?

**Comment by Borda** [comment]
> @williamFalcon how do I go about adding the example to the colab notebook?

We had a discussion with @ethanwharris and @MattPainter01 some time ago and we agreed to have it as a notebook in this repo which is connected Collab on request and also used as an example in Docs, right?

**Comment by djbyrne** [comment]
The circleci tests seem to be failing due to using typing OrderedDict. Is there any reason why this should be failing?

File: CHANGELOG.md [hunk 4]
```diff
@@ -8,6 +8,7 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 
 ### Added
 
+- Added dqn lightning example ([#1232](https://github.com/PyTorchLightning/pytorch-lightning/pull/1232))
```
Comment (Borda) [inline CHANGELOG.md]: could you five a full name of the example and then shortcut in brackets
```suggestion
- Added Reinforcement Learning - Deep Q-network (DQN) lightning example ([#1232](https://github.com/PyTorchLightning/pytorch-lightning/pull/1232))
```

**Comment by Borda** [comment]
@djbyrne could you rebase master, seems that you are missing the recent test/example split

**Comment by djbyrne** [comment]
> @djbyrne could you rebase master, seems that you are missing the recent test/example split

I rebased with master, but it seems that the ubuntu and osx tests fails when uploading the pytest results. Any ideas why?

**Comment by williamFalcon** [comment]
@djbyrne just restarted jobs. working to merge this ASAP :)

**Comment by Borda** [comment]
@djbyrne pls next time did rebase, now it seems like you did merge since it shows 67 changed files...
but  GREAT thanks for the RL example :zap:

**Pull request merged**